### ISSUE TITLE
Properties in layer per cell vs. tileset properties

### ISSUE THREAD

**Issue by biouxtai** [mapeditor/tiled#31]
I would like to see a feature for having properties based on a cell in a layer rather than base it on the tileset.  Tileset is too global.  If I have a small sample of tiles in a map that need one specific property (object area is not feasible) then creating a new tileset to handle this doesn't make sense.  Especially if there are several permutations of the property needed with the same tile. Custom properties on the cell in the layer would be much better.

**Comment by twistedjoe** [comment]
I know it's old, but tiled is useless on my usecase without this feature and I don't know where to upvote this feature. So here is my bump. 

Tried @biouxtai but it is unstable on my side (it crash a lot on osx). 
I will try to merge with the new codebase to check if it fix the issues.

**Comment by twistedjoe** [comment]
Ported to the current version here: https://github.com/samueldr/tiled/tree/feature/cellproperties

Added lua export, still missing the others since we only use the lua one.

Also still missing a way to tell which cell have properties. I don't really need this feature, but I can see how it would be useful.

**Comment by lordnoriyuki** [comment]
Just chiming in here saying that I really would love the functionality from the fork to land sometime.

**Comment by bjorn** [comment]
@twistedjoe Would you be interested in opening a pull request for your changes? This feature is still requested occasionally and a pull request would make it easier for me to review the changes and provide feedback.

**Comment by twistedjoe** [comment]
Sure! I think it would still need some "visual feedback", but it does the trick for us.

**Comment by twistedjoe** [comment]
There you go. 
https://github.com/bjorn/tiled/pull/1139

**Comment by twistedjoe** [comment]
We were working on a quite old version. So this might require some adjustment.

**Comment by bjorn** [comment]
@twistedjoe Thanks!

**Comment by Phlosioneer** [comment]
This is an old thread but a lot of people seem to want this kind of property. It's possible with a little bit of creativity to solve it, and old games (old pokemon games, snes titles, and others) had clever solutions.

First, a use case. Suppose you have a map of a dungeon. You need to assign certain floor tiles to be pitfall traps; you need some walls to be destructible; you need some tiles to be light sources; and you'd like to make most tiles have a "encounter chance" between 0% and 100%.

Here's what you can do. First, for each property, create a new tile layer. You can save space by making properties share a layer if they're mutually exclusive; in our case, we can store pitfall traps (only on floor tiles) and destructible walls (only on wall tiles) in one layer. Then we have a light-sources layer, and we have an encounter-chance layer.

Next, make a new tileset. Make tiles with english words, letters, simple colors, or simple shapes - they're only visible to you, the level designer, so they can be whatever is easiest to draw. In our case, we're going to make a tile labeled "T" for trap, a tile labeled "D" for destructible, a tile labeled "L" for lightsource, and tiles labeled "10%", "20%", "30%", "40%", "50%", "60%", "70%", "80%", "90%", and "100%".

Okay. Now back to Tiled. Create a new tileset file using the new image, then use the tile properties as needed. In this case, we'll assign the numbers 0.1, 0.2, 0.3, ... to the tiles named "10%", "20%", "30%", etc, using a custom property. That'll make working with these tiles in code a little bit nicer.

Go back to your map file. On the trap layer, place the new tile layers above the original dungeon layer. For each new layer, set the opacity to about 50%, and set the type field to something useful like "metadata".

On the trap layer, use the T tile wherever you want a pit trap, and a D tile wherever you want a destructible wall. Continue over to the light-source layer, and place an L tile wherever you want a light source. On the encounter-chance layer, use one of the probability tiles (10% - 100%) to assign probabilities throughout the dungeon.

Finally, in your game code, add an additional rule to never draw a layer if its type is "metadata".

This technique of metadata-tiles is very old - the oldest use I'm aware of is Pokemon Gold/Silver, where "cutscene" triggers are coded using invisible tiles, among other things.

**Comment by Phlosioneer** [comment]
That said, it does give a hint about how we could implement this directly in tiled.

We can setup a "virtual" tile-layer per property added to tiles. In the UI, it'll feel like selecting a tile and changing properties, particularly something like "painting" a property using existing tile placement tools. Indicating the presence of properties can be just a semi-opaque first-letter-of-property-name thing.

I think that it's very important to keep the feature-set of tile properties and the feature-set of object layers as separate as possible.

**Comment by AJLProjects** [comment]
@bjorn just to add to this discussion I recently encountered an issue where I wanted to place a chest tile, and it would be nice to customize the contents of the chest using a text/json.

I have been told by eishiya that I should use a object. The problem is that is bad for a couple of different reasons: The position may be off by a bit so may round down to the wrong position. You may say turn snapping on but then you are 'emulating' the grid by doing so, which is philosophically wrong. Also when you save as an object, its in a separate part of the saved datastructure that is processed differently thnt tiles on most game logic code. For example I might have a switch satement for rendering tiles, and there may be another switch statement for rendering enemies that are objects. The problem is that you will have chest in amongst the enemies, which doesnt make sense really.

It would be better to have another dictionary from grid positon => text/json content, that would show up if there is at least one object with tile metadata/properties.

**Comment by Phlosioneer** [comment]
@AJLProjects That is exactly the use case for objects. Which would you rather 1) have an entire separate parser specification, custom parsing code, that is fixed in place, or 2) call `round()` on the x and y coordinates, with the flexibility to move it to a different location if your level layout changes?

if you really want to have the tile as a chest, you can add a spriteless object on top with the data. 

If this feels messy, welcome to game development. Every map and video game is cobbled together with shortcuts and almost-aligned things.

**Comment by AJLProjects** [comment]
@Phlosioneer NO. Objects are for free floating objects like moving entities e.g. mobs. 
If it needs to conform to the grid and is a grid tile, then it should remain as a tile. 
Stop the mental gymnastics, thinking in terms of what already exists.

**Comment by Phlosioneer** [comment]
@AJLProjects Hate to burst your bubble, but… in Minecraft, chests are a tile + an entity, and an object connecting the two. Minecraft uses the same method of entity-on-top-of-tile data storage. And every animated entity works the same, along with anything that spans voxels. Doors are an entities, furnaces are entities, sand and gravel are temporary entities, beds are entities (they have a block for each half, bout those models are hidden and a replacement entity is visible, or else the lighting wouldn’t work), armor stands are entities, frames are entities, paintings are entities, fire flames are each entities… the list goes on. 

You can make a mod that gives chests health and allow them to take knockback damage every time they close. There are some performance related restrictions on tile entities but generally anything goes.

**Comment by Phlosioneer** [comment]
Unity engine works similarly, when a tile system is involved; tiles can’t have data.

**Comment by AJLProjects** [comment]
The fact this is even a contentious feature means it should be implemented, because it really boils down to the developers workflow, and their individual usecase. For me it would be much easier to place the tile and annotate the tile with data. This is personal preference, the same way in photoshop there are a million ways to do the same thing, because its accomodating to personal preference and the needs of the project. The needs of **my** project prefer a different setup to what exists currently.
This is a general purpose tool and should be accomodating and not opinionated. Even if you dissagree with my workflow / game implementation.
